File: ios/LanGaming/Sources/LanGaming/LanGamingHost.swift
import Foundation

/// HTTP host on `locxxLanPort` using the same wire frames as BLE (`ProtocolCodec` / docs/protocol.md).
public final class LanGamingHost {
    private let listener: LanGamingListener
    private var http: LanHttpServer?
    private let sessionNonce = Data((0 ..< 16).map { _ in UInt8.random(in: 0 ... 255) })
    private var nextPlayerId = 1
    private var tokenToPeer: [String: Peer] = [:]
    private let sync = NSLock()

    private final class Peer {
        let token: String
        let playerId: Int
        let displayName: String
        private let cond = NSCondition()
        private var queue: [Data] = []

        init(token: String, playerId: Int, displayName: String) {
            self.token = token
            self.playerId = playerId
            self.displayName = displayName
        }

        func take(timeout: TimeInterval) -> Data? {
            cond.lock()
            defer { cond.unlock() }
            let deadline = Date().addingTimeInterval(timeout)
            while queue.isEmpty {
                if !cond.wait(until: deadline) {
                    return nil
                }
            }
            return queue.removeFirst()
        }

        func offer(_ data: Data) {
            cond.lock()
            queue.append(data)
            cond.signal()
            cond.unlock()
        }
    }

    public init(listener: LanGamingListener) {
        self.listener = listener
    }

    public func sessionNonceData() -> Data { sessionNonce }

    public func start() {
        let server = LanHttpServer(port: locxxLanPort) { [weak self] req in
            guard let self else {
                return (404, ["Content-Type": "text/plain"], Data("gone".utf8))
            }
            return self.handle(req: req)
        }
        http = server
        server.start()
    }

    public func stop() {
        http?.stop()
        http = nil
        sync.lock()
        tokenToPeer.removeAll()
        nextPlayerId = 1
        sync.unlock()
    }

    public func broadcast(_ frame: Data) {
        sync.lock()
        let peers = Array(tokenToPeer.values)
        sync.unlock()
        let bytes = Data(frame)
        for p in peers {
            p.offer(Data(bytes))
        }
    }

    public func sendToPeer(addressToken: String, frame: Data) {
        sync.lock()
        let peer = tokenToPeer[addressToken]
        sync.unlock()
        peer?.offer(Data(frame))
    }

    private func handle(req: LanHttpRequest) -> (Int, [String: String], Data) {
        switch (req.method, req.path) {
        case ("POST", "/locxx/v1/hello"):
            return serveHello(body: req.body)
        case ("POST", "/locxx/v1/send"):
            return serveSend(token: req.query["token"] ?? "", body: req.body)
        case ("GET", "/locxx/v1/poll"):
            return servePoll(token: req.query["token"] ?? "")
        default:
            return (404, ["Content-Type": "text/plain"], Data("not found".utf8))
        }
    }

    private func serveHello(body: Data) -> (Int, [String: String], Data) {
        let decoded: DecodedFrame
        do {
            decoded = try ProtocolCodec.decodeFrame(body)
        } catch {
            mainError("decode")
            return (400, [:], Data())
        }
        guard decoded.messageType == WireMessageType.clientHello else {
            return (400, ["Content-Type": "text/plain"], Data("expected hello".utf8))
        }
        let hello: ClientHelloPayload
        do {
            hello = try ProtocolCodec.parseClientHello(decoded.payload)
        } catch {
            mainError("hello")
            return (400, [:], Data())
        }
        guard hello.protocolVersion == wireVersion else {
            mainError("protocol mismatch")
            return (400, [:], Data())
        }
        sync.lock()
        guard nextPlayerId <= 8 else {
            sync.unlock()
            mainError("room full")
            return (503, ["Content-Type": "text/plain"], Data("room full".utf8))
        }
        let pid = nextPlayerId
        nextPlayerId += 1
        let token = UUID().uuidString
        tokenToPeer[token] = Peer(token: token, playerId: pid, displayName: hello.displayName)
        sync.unlock()

        let welcome: Data
        do {
            welcome = try ProtocolCodec.buildHostWelcome(
                protocolVersion: wireVersion,
                playerId: UInt8(pid),
                sessionNonce: sessionNonce
            )
        } catch {
            sync.lock()
            tokenToPeer.removeValue(forKey: token)
            sync.unlock()
            return (500, [:], Data())
        }
        DispatchQueue.main.async {
            self.listener.onPeerConnected(address: token, playerId: pid, displayName: hello.displayName)
        }
        let headers: [String: String] = [
            "Content-Type": "application/octet-stream",
            locxxHttpTokenHeader: token,
        ]
        return (200, headers, welcome)
    }

    private func serveSend(token: String, body: Data) -> (Int, [String: String], Data) {
        sync.lock()
        let peer = tokenToPeer[token]
        sync.unlock()
        guard let peer else {
            return (404, [:], Data())
        }
        let decoded: DecodedFrame
        do {
            decoded = try ProtocolCodec.decodeFrame(body)
        } catch {
            mainError("decode")
            return (400, [:], Data())
        }
        switch decoded.messageType {
        case WireMessageType.ping:
            peer.offer(ProtocolCodec.encodeFrame(messageType: WireMessageType.pong))
        default:
            DispatchQueue.main.async {
                self.listener.onMessageReceived(address: token, messageType: decoded.messageType, payload: decoded.payload)
            }
        }
        return (204, [:], Data())
    }

    private func servePoll(token: String) -> (Int, [String: String], Data) {
        sync.lock()
        let peer = tokenToPeer[token]
        sync.unlock()
        guard let peer else {
            return (404, [:], Data())
        }
        if let frame = peer.take(timeout: 55) {
            return (200, ["Content-Type": "application/octet-stream"], frame)
        }
        return (204, [:], Data())
    }

    private func mainError(_ message: String) {
        DispatchQueue.main.async {
            self.listener.onError(message)
        }
    }
}

File: android/lan-gaming/src/main/kotlin/com/locxx/langaming/LanGamingHost.kt
package com.locxx.langaming

import android.os.Handler
import android.os.Looper
import fi.iki.elonen.NanoHTTPD
import java.io.ByteArrayInputStream
import java.nio.charset.StandardCharsets
import java.util.UUID
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

const val LOCXX_HTTP_TOKEN_HEADER = "X-LocXX-Token"

/** Socket read timeout (ms) passed to [NanoHTTPD.start]; same default as NanoHTTPD (5 min). */
private const val LOCXX_NANOHTTPD_READ_TIMEOUT_MS = 5 * 60 * 1000

/**
 * HTTP host on [LOCXX_LAN_PORT]: handshake and long-poll transport for the same wire frames as BLE
 * (see [ProtocolCodec] / docs/protocol.md).
 */
class LanGamingHost(
    private val listener: LanGamingListener
) : NanoHTTPD(LOCXX_LAN_PORT) {

    /**
     * Never gzip responses. Android’s [HttpURLConnection] often sends Accept-Encoding: gzip; NanoHTTPD
     * would then use chunked Transfer-Encoding + [GZIPOutputStream] even for short `text/plain` bodies
     * (e.g. long-poll 204). If the client closes the socket first (timeout, app backgrounded, network
     * change), writing gzip headers fails with [SocketException] and spams “Could not send response”.
     * Binary game frames are tiny; gzip is unnecessary here.
     */
    override fun useGzipWhenAccepted(r: Response): Boolean = false

    /** Shown to joining players before they connect ([serveSessionInfo]). */
    @Volatile
    var advertisedHostDisplayName: String = ""

    private val mainHandler = Handler(Looper.getMainLooper())
    private val sessionNonce: ByteArray = ByteArray(16).also { java.security.SecureRandom().nextBytes(it) }
    private val nextPlayerId = AtomicInteger(1)
    private val tokenToPeer = ConcurrentHashMap<String, Peer>()

    private data class Peer(
        val token: String,
        val playerId: Int,
        val displayName: String,
        val outbound: LinkedBlockingQueue<ByteArray> = LinkedBlockingQueue()
    )

    fun sessionNonceBytes(): ByteArray = sessionNonce.copyOf()

    fun beginHosting(): Boolean =
        try {
            start(LOCXX_NANOHTTPD_READ_TIMEOUT_MS, false)
            true
        } catch (e: Exception) {
            listener.onErrorUi("start: ${e.message ?: "failed"}")
            false
        }

    fun endHosting() {
        stop()
        tokenToPeer.clear()
        nextPlayerId.set(1)
    }

    fun broadcast(frame: ByteArray) {
        val data = frame.copyOf()
        for (p in tokenToPeer.values) {
            p.outbound.offer(data.copyOf())
        }
    }

    fun sendToPeer(addressToken: String, frame: ByteArray) {
        tokenToPeer[addressToken]?.outbound?.offer(frame.copyOf())
    }

    override fun serve(session: IHTTPSession): Response {
        val path = session.uri.substringBefore('?').let { p ->
            if (p.startsWith("/")) p else "/$p"
        }
        return try {
            when (session.method) {
                Method.POST -> when (path) {
                    "/locxx/v1/hello" -> serveHello(session)
                    "/locxx/v1/send" -> serveSend(session)
                    else -> notFound()
                }
                Method.GET -> when (path) {
                    "/locxx/v1/poll" -> servePoll(session)
                    "/locxx/v1/session_info" -> serveSessionInfo()
                    else -> notFound()
                }
                else -> notFound()
            }
        } catch (e: Exception) {
            newFixedLengthResponse(
                Response.Status.INTERNAL_ERROR,
                NanoHTTPD.MIME_PLAINTEXT,
                e.message ?: "error"
            )
        }
    }

    private fun serveHello(session: IHTTPSession): Response {
        val body = readRequestBody(session)
        val decoded = ProtocolCodec.decodeFrame(body).getOrElse {
            return text(Response.Status.BAD_REQUEST, it.message ?: "frame")
        }
        if (decoded.messageType != WireMessageType.CLIENT_HELLO) {
            return text(Response.Status.BAD_REQUEST, "expected hello")
        }
        val hello = ProtocolCodec.parseClientHello(decoded.payload).getOrElse {
            return text(Response.Status.BAD_REQUEST, it.message ?: "hello")
        }
        if (hello.protocolVersion != WIRE_VERSION) {
            return text(Response.Status.BAD_REQUEST, "protocol mismatch")
        }
        val pid = nextPlayerId.getAndIncrement()
        if (pid > 8) {
            nextPlayerId.decrementAndGet()
            return text(Response.Status.BAD_REQUEST, "room full")
        }
        val token = UUID.randomUUID().toString()
        tokenToPeer[token] = Peer(token = token, playerId = pid, displayName = hello.displayName)
        val welcome = ProtocolCodec.buildHostWelcome(WIRE_VERSION, pid.toByte(), sessionNonce)
        listener.onPeerConnectedUi(token, pid, hello.displayName)
        return newFixedLengthResponse(
            Response.Status.OK,
            "application/octet-stream",
            ByteArrayInputStream(welcome),
            welcome.size.toLong()
        ).apply { addHeader(LOCXX_HTTP_TOKEN_HEADER, token) }
    }

    private fun serveSend(session: IHTTPSession): Response {
        val token = session.parameters["token"]?.firstOrNull()
            ?: return text(Response.Status.BAD_REQUEST, "token")
        val peer = tokenToPeer[token] ?: return text(Response.Status.NOT_FOUND, "unknown token")
        val body = readRequestBody(session)
        val decoded = ProtocolCodec.decodeFrame(body).getOrElse {
            listener.onErrorUi(it.message ?: "decode")
            return text(Response.Status.BAD_REQUEST, "decode")
        }
        when (decoded.messageType) {
            WireMessageType.PING -> {
                peer.outbound.offer(ProtocolCodec.encodeFrame(WireMessageType.PONG))
            }
            else -> {
                listener.onMessageReceivedUi(token, decoded.messageType, decoded.payload)
            }
        }
        return newFixedLengthResponse(Response.Status.NO_CONTENT, NanoHTTPD.MIME_PLAINTEXT, "")
    }

    private fun servePoll(session: IHTTPSession): Response {
        val token = session.parameters["token"]?.firstOrNull()
            ?: return text(Response.Status.BAD_REQUEST, "token")
        val peer = tokenToPeer[token] ?: return text(Response.Status.NOT_FOUND, "unknown token")
        val frame = peer.outbound.poll(55, TimeUnit.SECONDS)
        return if (frame == null) {
            newFixedLengthResponse(Response.Status.NO_CONTENT, NanoHTTPD.MIME_PLAINTEXT, "")
        } else {
            newFixedLengthResponse(
                Response.Status.OK,
                "application/octet-stream",
                ByteArrayInputStream(frame),
                frame.size.toLong()
            )
        }
    }

    private fun notFound(): Response = text(Response.Status.NOT_FOUND, "not found")

    private fun serveSessionInfo(): Response {
        val json = JSONObject().apply {
            put("hostDisplayName", advertisedHostDisplayName)
        }
        val bytes = json.toString().toByteArray(StandardCharsets.UTF_8)
        return newFixedLengthResponse(
            Response.Status.OK,
            "application/json; charset=utf-8",
            ByteArrayInputStream(bytes),
            bytes.size.toLong()
        )
    }

    private fun text(status: Response.Status, msg: String): Response =
        newFixedLengthResponse(status, NanoHTTPD.MIME_PLAINTEXT, msg)

    private fun readRequestBody(session: IHTTPSession): ByteArray {
        val rawLen = session.headers["content-length"] ?: return ByteArray(0)
        val len = rawLen.toIntOrNull() ?: 0
        if (len <= 0) return ByteArray(0)
        val buf = ByteArray(len)
        var off = 0
        val input = session.inputStream
        while (off < len) {
            val n = input.read(buf, off, len - off)
            if (n <= 0) break
            off += n
        }
        return if (off == len) buf else buf.copyOf(off)
    }

    private fun LanGamingListener.onPeerConnectedUi(token: String, pid: Int, name: String) {
        mainHandler.post { onPeerConnected(token, pid, name) }
    }

    private fun LanGamingListener.onMessageReceivedUi(token: String, type: Byte, payload: ByteArray) {
        mainHandler.post { onMessageReceived(token, type, payload) }
    }

    private fun LanGamingListener.onErrorUi(msg: String) {
        mainHandler.post { onError(msg) }
    }
}
